### DISCUSSION TITLE
Overloading on return type

### DISCUSSION THREAD

**Discussion by julealgon** [dotnet/csharplang#6638] [Language Ideas]
What are the current constraints that prevent overloading on the return type alone?

Being able to overload on the return type could open the door for expanding on existing APIs in a very natural way. Say for example the Discriminated Union proposal goes through:
- https://github.com/dotnet/csharplang/issues/113

One could then overload the `Parse` methods to also work with something like `Optional<T>`.

Similarly, we could overload something like `ArgumentNullException.ThrowIfNull` to return the validated value back as per this:
- https://github.com/dotnet/runtime/issues/70515

Apologies if such discussion/request already exists (I did my best searching for it and didn't find anything).

**Comment by HaloFour** [comment]
I'd wager complexity and ambiguity.  Overload resolution is already the most complicated thing that the compiler does, probably by an order of magnitude.  To add the return type to the mix would make that a lot worse as not only does it need to resolve based on inputs but also based on outputs.  I also don't see how it could remotely work with variable inference as, given your example, there's no way the compiler could infer which overload to call given the following:

```cs
var x = int.Parse(s);
```

I believe that if the runtime were to consider overloads for the purposes of `Option<T>` they would do so for `TryParse`, not `Parse`, and that would not require any new form of overloading:

```cs
public struct Int32 {
    public static bool TryParse(string input, out int result);
    public static Option<T> TryParse(string input);
}
```

**Reply by julealgon** [reply]
> I'd wager complexity and ambiguity. Overload resolution is already the most complicated thing that the compiler does, probably by an order of magnitude. To add the return type to the mix would make that a lot worse as not only does it need to resolve based on inputs but also based on outputs.

I see. So it's not that "it's not possible" per se, but just very hard to get right/complex/etc.

> I also don't see how it could remotely work with variable inference as, given your example, there's no way the compiler could infer which overload to call given the following:
> 
> ```cs
> var x = int.Parse(s);
> ```

Oh for sure, this wouldn't "just work" today.

However, what if:
1. We had some control over a "preferable" overload. Similar to how attributes exist to select "the preferred dependency injection constructor"? You could then override the "decision" by typing the variable explicitly, otherwise it would fall on the "default" overload (something like that)
2. What if `var` inference was "postponed" until the compiler could see where you are using it?

Imagine a scenario like this:
```csharp
var x = int.Parse(s);
MyLocalFunction(value);

void MyLocalFunction(int value) => ...
```

In the example above, the type of `x` could be inferred as `int` even if `int.Parse` had multiple overloads with different return types.

Would something like that be feasible? I assume so.

> I believe that if the runtime were to consider overloads for the purposes of `Option<T>` they would do so for `TryParse`, not `Parse`, and that would not require any new form of overloading:
> 
> ```cs
> public struct Int32 {
>     public static bool TryParse(string input, out int result);
>     public static Option<T> TryParse(string input);
> }
> ```

Fair enough.

---

Also, I could see this being especially interesting for overloading a `void`-returning method... This could even be a special case as it wouldn't require complex rules to disambiguate.

Say we had:
```csharp
void ThrowIfNull<T>(T value)...
T ThrowIfNull<T>(T value)...
```

Then you could infer which method is used by the presence of an assignment:
```csharp
ThrowIfNull(value); // picks first
var anotherValue = ThrowIfNull(value); // picks second
_ = ThrowIfNull(value); // explicitly picks second as well
```

**Reply by HaloFour** [reply]
> I see. So it's not that "it's not possible" per se, but just very hard to get right/complex/etc.

Right, it would require an enormous amount of work and testing, all to avoid having to use a different identifier.

> We had some control over a "preferable" overload.

That just complicates it more, now the compiler has to figure out how to evaluate multiple candidates based on both the priority it assumes itself given type coercion and the possibility of these attributes being (or not being) applied to combinations of those overloads.

It's been mentioned that if the C# team could start over today they might choose the Go route and just forbid overloads, that's how complicated it is.  😄 

> What if `var` inference was "postponed" until the compiler could see where you are using it?

Much more complicated, especially when you could have multiple assignments and usages of any given variable.

**Comment by Joe4evr** [comment]
Tangentially: There was some talk about allowing return-type overloading [in the context of deprecating certain APIs](https://github.com/dotnet/runtime/issues/65965). It's not mentioned in the issue itself, but I recall during the review it was mentioned that such a "binary-compat only" flag would have the most appeal if it allowed return-type overloading in conjunction.

It's still massively complicated to implement, tho.
